File: benchmarks/tasks/type_definition/type_definition.cpp
// SPDX-License-Identifier: MIT-0

#include <iostream>
#include <memory>

struct List {
    int head;
    std::shared_ptr<List> tail;
    List(int h, std::shared_ptr<List> t) : head(h), tail(t) {}
};

using ListPtr = std::shared_ptr<List>;

ListPtr nil() { return nullptr; }
ListPtr cons(int head, ListPtr tail) { return std::make_shared<List>(head, tail); }

int length(ListPtr xs) {
    if (!xs) return 0;
    return 1 + length(xs->tail);
}

ListPtr append(ListPtr xs, int x) {
    if (!xs) return cons(x, nil());
    return cons(xs->head, append(xs->tail, x));
}

int main() {
    auto xs = cons(1, cons(2, cons(3, nil())));
    auto ys = append(xs, 4);
    std::cout << length(ys) << std::endl;
    return 0;
}

File: benchmarks/tasks/type_definition/type_definition.hs
-- SPDX-License-Identifier: MIT-0

data List a = Nil | Cons a (List a)

len :: List a -> Int
len Nil        = 0
len (Cons _ t) = 1 + len t

append :: List a -> a -> List a
append Nil x        = Cons x Nil
append (Cons h t) x = Cons h (append t x)

main :: IO ()
main =
  let xs = Cons 1 (Cons 2 (Cons 3 Nil))
      ys = append xs 4
  in print (len ys)
